test/libsolidity/syntaxTests/dataLocations/data_location_in_function_type.sol

```diff
@@ -0,0 +1,4 @@
+library L {
+    struct Nested { uint y; }
+    function c(function(Nested memory) external returns (uint)[] storage) external pure {}
```

Comment:
If we feel paranoid, we could add tests for arguments of a function type, which in turn has arguments or return values of a function type that has arguments that require data locations, but judging from the code it should work fine.